lib/addons/velero/index.ts

```diff
@@ -0,0 +1,259 @@
+            values: {
+                configuration: {
+                    backupStorageLocation: {
+                        prefix: props.values.configuration.backupStorageLocation.prefix ?? "velero/" + cluster.clusterName,
+                        bucket: bucketName,
+                        config:{
+                           region: props.values.configuration.backupStorageLocation.config.region ?? cluster.stack.region,
+                        }
+                    },
+                    volumeSnapshotLocation:{
+                        config:{
+                            region: props.values.configuration.backupStorageLocation.config.region ?? cluster.stack.region
+                        }
+                    }
+                },
+                // IAM role for Service Account
+                serviceAccount: {
+                    server: {
+                        create: false,
+                        name: veleroServiceAccountName,    
+                    }
+                }             
+            }
+        };
+
+        const values = merge(props.values, valueVariable.values) ?? {}; 
+ 
+        this.chartNode = cluster.addHelmChart("velero-addon", {
+            chart: "velero",
+            repository: "https://vmware-tanzu.github.io/helm-charts/",
+            release: Constants.SSP_ADDON,
+            namespace: veleroNamespace,
+            version: props.version,
+            values: values
+        });
+        return this.chartNode;
+    }
+
+    /**
+     * Return S3 Bucket
+     * @param clusterInfo 
+     * @param id S3-Bucket-Postfix 
+     * @param existingBucketName exiting provided S3 BucketName if it exists 
+     * @returns the existing provided S3 bucket name or the newly created S3 bucket name
+     */
+    protected returnS3BucketName(clusterInfo: ClusterInfo, id: string, existingBucketName: null|string ): string {
+        if (!existingBucketName){
+            const bucket = new s3.Bucket(clusterInfo.cluster, "velero-${id}", {
+                encryption: s3.BucketEncryption.KMS_MANAGED, // Velero Known bug for support with S3 with SSE-KMS with CMK, thus it does not support S3 Bucket Key: https://github.com/vmware-tanzu/helm-charts/issues/83
+                blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL, // Block Public Access for S3
+                publicReadAccess: false,
+                enforceSSL: true // Encryption in Transit
+            });
+            return bucket.bucketName;
+        }
+        else {
+            return existingBucketName;
+        }
+    }
+
+    /**
+     * Return Velero Namespace where Velero will be installed onto
+     * @param clusterInfo
+     * @param defaultName the Default Namespace for Velero if nothing specified 
+     * @param namespace
+     * @returns the namespace created or existed.
+     */
+    protected createNamespaceIfNeeded(clusterInfo: ClusterInfo, defaultName: string, namespace?: {name: string, create: boolean}): string {
```

Comment:
If namespace is created, you will need the actual namespace node, so that you can create proper dependencies. For example, you cannot create service account until namespace is created.  See `sa.node.addDependency(namespace);` in ArgoCD add-on.

Reply:
I referred to ArgoCD add-on, but in ArgoCD a namespace has to be created. But in Velero's case, it supports an existing namespace. I have commented in the actual code and please have a look.

Reply:
I understand your comment. However, in case namespace is created lack of dependency between service account and namespace WILL cause sporadic failures. 
It means that the code must create a CFN dependency between namespace and service account *if namespace is created by us*.

You can make this function do this:

```typescript

protected createNamespaceIfNeeded( ... ) : { name: string, manifest?: KubnernetesManifest { 
...
    return {name: somename, manifest: (either manifest or undefined)
}
```

Then in deploy (pseudocode):

```typescript

const namespace = createNamespaceIfNeeded(...);
sa = createServiceAccount(...);

if(namespace.manifest) {
   sa.node.addDependency(namespace.manifest);
}

```

Reply:
Thanks for the detailed code recommendations, it worked perfectly. Much appreciated for your patience, I learnt a lot from this process.